Question 1a: 
Given the string 'great', use an appropriate index command that returns 'a'
from the string. Given s = ‘great’. (0.5 point)
Expected output: ‘a’

In [60]:
s = 'great'
print("Expected output: " + s[3])

Expected output: a


Question 1b:
Write a program to find out the second smallest number in the list given
below. (0.5 point)
li=[32, 3, 5, 8, 33, 7] & Expected output: 5

In [61]:
li=[32, 3, 5, 8, 33, 7]
li.sort()
print(li[1])

5


Question 1c: Find out the unique values of the list given below. (0.5 point)
mylist=[1,1,5,3,3,6,11,11] & Expected output: [1,5,3,6,11]

In [62]:
mylist=[1,1,5,3,3,6,11,11]
uniquelist = []
for n in mylist:
    if n not in uniquelist:
        uniquelist.append(n)
print(uniquelist)

[1, 5, 3, 6, 11]


Question 1d:
Write a program to print cube root of 216 using appropriate syntax. (0.5
point)
Expected output: 6

In [63]:
import numpy as np
n = 216
croot = np.cbrt(n)
print(croot)

6.0


Question 2:
 Replace the value 'hello' in the below nested list with 'bye'. (1 point)
 a = [1, 2, [3, 4, 'hello']] & Expected output: [1, 2, [3, 4, 'bye']]

In [64]:
a = [1, 2, [3, 4, 'hello']]
a[2][2] = 'bye'
print(a)

[1, 2, [3, 4, 'bye']]


Question 3:
Given the below dictionary, write a program that returns [‘object0’]. (1 point)
dict = {'a':9, 'b':(2,3,1), 'c':['object0','object1','object2']}
Expected output: [‘object0’]

In [65]:
dict = {'a':9, 'b':(2,3,1), 'c':['object0','object1','object2']}
answer = dict['c'][0]
print(answer)

object0


Question 4:
Write a program to calculate the area of a circle. The program should ask for an
appropriate input and print the computed area. (2 points)
a: Input text: Please enter the radius of the circle.
b. If the user enters a value of 5. The output should be: The area of the circle
is 78.54
c. This program should indicate that the input is invalid if the user enters a
character instead of a number as input. For instance if the user enters a
value of ‘a’ instead of 5 in the previous example the program should
prompt: Invalid input, please enter a number: _

In [67]:
import math
print("Please enter the radius of the circle. ")
while True:
    try:
        radius = float(input())
        break
    except ValueError:
        radius = print("Invalid input, please enter a number: _")


print("The area of the circle is " + str(round(math.pi * (radius**2),2)))

Please enter the radius of the circle. 
h
Invalid input, please enter a number: _
5
The area of the circle is 78.54


Question 5:
Write a program which will take two lists X & Y as input and it will return a list
having all the odd numbers present in the two lists. (Note:- common elements
can be repeated) (2 points)
Input list: X=[1,2,3,4,5] & Y=[5,6,7,8,9]
Expected Output: [1,3,5,5,7,9] or [1,3,5,7,9] both are accepted.

In [68]:
def oddnum(X,Y):
    odds = []
    X.extend(Y)
    for x in X:
        if x%2 != 0:
            odds.append(x)
    return odds

X=[1,2,3,4,5]
Y=[5,6,7,8,9]
odds_list = oddnum(X,Y)
print(odds_list)

[1, 3, 5, 5, 7, 9]


Question 6:
Write a function that prints the integers from 1 to 15. But for multiples of three
print "GREAT" instead of the number, and for the multiples of five print
"LEARNING". For numbers which are multiples of both three and five print
"GREATLEARNING". (2 points)

In [69]:
def foobar():
    for i in range(1,16):
        if i%3 == 0 and i%5 == 0:
            print("GREATLEARNING")
        elif i%3 == 0:
            print("GREAT")
        elif i%5 == 0:
            print("LEARNING")
        else:
            print(i)
foobar()

1
2
GREAT
4
LEARNING
GREAT
7
8
GREAT
LEARNING
11
GREAT
13
14
GREATLEARNING


# Part 2

##### 1. Import necessary libraries and read the provided dataset (online_sales.csv) and check the top 5 and random 5 samples of the dataframe. (0.5 point)

In [57]:
import pandas as pd
import numpy as np
df = pd.read_csv("online_sales.csv")
print(df.head(5))

  InvoiceNo StockCode                          Description  Quantity  \
0    536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1    536365     71053                  WHITE METAL LANTERN         6   
2    536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3    536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4    536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

    InvoiceDate  UnitPrice  CustomerID         Country  
0  12/1/10 8:26       2.55     17850.0  United Kingdom  
1  12/1/10 8:26       3.39     17850.0  United Kingdom  
2  12/1/10 8:26       2.75     17850.0  United Kingdom  
3  12/1/10 8:26       3.39     17850.0  United Kingdom  
4  12/1/10 8:26       3.39     17850.0  United Kingdom  


In [58]:
print(df.sample(5))

       InvoiceNo StockCode                    Description  Quantity  \
150491    549440     22570       FELTCRAFT CUSHION RABBIT         4   
10534     537237     22501     PICNIC BASKET WICKER LARGE         1   
43615     540098     22276  WASH BAG VINTAGE ROSE PAISLEY         1   
182318    552537    35598B      BLACK CHRISTMAS TREE 60CM         2   
77935     542794     20719         WOODLAND CHARLOTTE BAG         1   

          InvoiceDate  UnitPrice  CustomerID         Country  
150491   4/8/11 12:42       3.75     17162.0  United Kingdom  
10534    12/6/10 9:58      21.23         NaN  United Kingdom  
43615    1/4/11 15:50       2.55     16241.0  United Kingdom  
182318  5/10/11 10:11       2.95     17315.0  United Kingdom  
77935    2/1/11 10:56       1.63         NaN  United Kingdom  


##### 2. Check info of the dataframe and write your observations. Comment on data types and shape of the dataset. (0.5 point)

In [59]:
print(df.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 240007 entries, 0 to 240006
Data columns (total 8 columns):
 #   Column       Non-Null Count   Dtype  
---  ------       --------------   -----  
 0   InvoiceNo    240007 non-null  object 
 1   StockCode    240007 non-null  object 
 2   Description  239106 non-null  object 
 3   Quantity     240007 non-null  int64  
 4   InvoiceDate  240007 non-null  object 
 5   UnitPrice    240007 non-null  float64
 6   CustomerID   172782 non-null  float64
 7   Country      240007 non-null  object 
dtypes: float64(2), int64(1), object(5)
memory usage: 14.6+ MB
None


Comments:
There are 2 float64 dtypes, 2 int64, and 5 objects
The shape of the dataset is of 8 columns and 240007 rows

##### 3. Check for null values and report the percentage of null values of each column. And drop the rows having null values in it. (1 point)

In [4]:
for column in df:
    sumofnan = df[column].isnull().sum()
    print("The percentage of null values in " + column + " = " + str((sumofnan/len(df.index))*100))
print("Dropping rows with null values...")
df = df.dropna()

The percentage of null values in InvoiceNo = 0.0
The percentage of null values in StockCode = 0.0
The percentage of null values in Description = 0.37540571733324446
The percentage of null values in Quantity = 0.0
The percentage of null values in InvoiceDate = 0.0
The percentage of null values in UnitPrice = 0.0
The percentage of null values in CustomerID = 28.009599720008165
The percentage of null values in Country = 0.0
Dropping rows with null values...


##### 4. Check the statistical summary of the dataset. (Use inbuilt functions) (0.5 point)

In [5]:
print(df.describe())

            Quantity      UnitPrice     CustomerID
count  172782.000000  172782.000000  172782.000000
mean       12.244835       3.793566   15274.819941
std       257.898235     101.069930    1725.093177
min    -74215.000000       0.000000   12346.000000
25%         2.000000       1.250000   13842.000000
50%         5.000000       1.950000   15132.000000
75%        12.000000       3.750000   16814.000000
max     74215.000000   38970.000000   18287.000000


##### 5. Drop the instances having quantity less than zero. (1 point) (Note:- refer to the ‘Quantity’ feature)

In [6]:
dataindex = df.index
i = 0
for num in df["Quantity"]:
    if num < 0 :
        df.drop(dataindex[i])
    else:
        i += 1
print("dropped instances with negative quantities")

dropped instances with negative quantities


##### 6. Check unique values of the country and report the name of the country that has the highest number of instances/rows. (0.5 point)

In [7]:
unique_countries = df["Country"].unique()
group_df = df.groupby("Country")
maxofgroup = int(group_df.size().max())

for country in unique_countries:
    currcount = group_df.get_group(country).count()["Country"]
    
    if currcount == maxofgroup:
        print("The country with highest number of instances is: " + country)


The country with highest number of instances is: United Kingdom


##### 7. Create a new column with the name as ‘sales’ having total sales. The total sales is defined as Quantity*UnitPrice. (1 point)

In [8]:
df["Sales"] = df["Quantity"] * df["UnitPrice"]
df.head(5)

InvoiceNo StockCode                          Description  Quantity  \
0    536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1    536365     71053                  WHITE METAL LANTERN         6   
2    536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3    536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4    536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

    InvoiceDate  UnitPrice  CustomerID         Country  Sales  
0  12/1/10 8:26       2.55     17850.0  United Kingdom  15.30  
1  12/1/10 8:26       3.39     17850.0  United Kingdom  20.34  
2  12/1/10 8:26       2.75     17850.0  United Kingdom  22.00  
3  12/1/10 8:26       3.39     17850.0  United Kingdom  20.34  
4  12/1/10 8:26       3.39     17850.0  United Kingdom  20.34

##### 8. Report the top 5 countries in terms of sales. (2 points)
##### a. Consider the size of sales.
##### b. Consider the mean value of sales.


In [16]:
import operator
sales = {}
mean_sales = {}
for country in unique_countries:
    maxsale = group_df.get_group(country).max()["Sales"]
    sales[country] = maxsale
print("Top 5 countries in size of sales : ")
highestsales = dict(sorted(sales.items(), key=operator.itemgetter(1), reverse=True)[:5])
print(highestsales.keys())


Top 5 countries in size of sales : 
dict_keys(['United Kingdom', 'Japan', 'Singapore', 'EIRE', 'Netherlands'])


In [15]:
mean_sales = {}
for country in unique_countries:
    meansale = group_df.get_group(country).mean()["Sales"]
    mean_sales[country] = meansale
print("Top 5 countries in mean value of sales : ")
highestmeans = dict(sorted(mean_sales.items(), key=operator.itemgetter(1), reverse=True)[:5])
print(highestmeans.keys())

Top 5 countries in mean value of sales : 
dict_keys(['Australia', 'Netherlands', 'Japan', 'Sweden', 'Denmark'])


C:\Users\shaur\AppData\Local\Temp/ipykernel_12728/3722523413.py:3: FutureWarning: Dropping of nuisance columns in DataFrame reductions (with 'numeric_only=None') is deprecated; in a future version this will raise TypeError.  Select only valid columns before calling the reduction.
  meansale = group_df.get_group(country).mean()["Sales"]


##### 9. Report the top 5 products which bring the highest sales. Use StockCode for product information. (1 point)

In [17]:
print("Top 5 products which bring in the highest sales: ")
print(df.nlargest(5, 'Sales')['StockCode'])

Top 5 products which bring in the highest sales: 
61619     23166
222680    22502
173382     POST
52711     21108
160546    21108
Name: StockCode, dtype: object


##### 10. Convert the ‘InvoiceDate’ into a date format and report the month on which the maximum sales occur? (2 points)

In [53]:
from datetime import datetime

month_sale = {}
dataindex = df.index
i = 0
for date in df["InvoiceDate"]:
    datetime_obj = datetime.strptime(date, '%m/%d/%y %H:%M')
    month = datetime_obj.strftime("%B")
    month_sale[month] = df['Sales'][dataindex[i]]
    i+=1

Keymax = max(month_sale, key= lambda x: month_sale[x])
print(Keymax)





May
